### PR TITLE
Convert paths registered to their realpath

### PR DESCRIPTION
**Pull request by navarr**
### Description (*)

Many Magento tools expect that the path given to the ComponentRegistrar be the real path of the directory.  As such, it is possible for extension developers to make mistakes that are difficult to debug but result in issues (such as a third party module having its Test classes compiled [BUNDLE-1798]

This change makes Magento a little less error-prone.

### Fixed Issues (if relevant)

1. BUNDLE-1798

### Manual testing scenarios (*)

N/A

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @navarr. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me $VERSION instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by navarr** [comment]
#squashtoberfest

**Comment by orlangur** [comment]
@navarr,

> Many Magento tools expect that the path given to the ComponentRegistrar be the real path of the directory

How it could happen if it was never true before?

**Comment by navarr** [comment]
@orlangur Maybe "real path" is the wrong phrase.  Let's say registered path is "blah/blah/../blah2" - it'll work for registration, but certain other tooling will fail (e.g. auto-generated regex patterns for excluding automated tests from DI compilation)

**Comment by sivaschenko** [comment]
Hi @navarr , thanks for your contribution! It would be great if you update the https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Component/Test/Unit/ComponentRegistrarTest.php to cover cases when realpath usage is necessary.

File: lib/internal/Magento/Framework/Component/ComponentRegistrar.php [hunk 1]
```diff
@@ -51,7 +51,7 @@ public static function register($type, $componentName, $path)
                 . 'has been already defined in \'' . self::$paths[$type][$componentName] . '\'.'
             );
         } else {
-            self::$paths[$type][$componentName] = str_replace('\\', '/', $path);
+            self::$paths[$type][$componentName] = realpath(str_replace('\\', '/', $path));
```
Comment (orlangur) [inline lib/internal/Magento/Framework/Component/ComponentRegistrar.php:54]: We must not operate with absolute file paths.

@navarr could you provide details of `BUNDLE-1798` please?

**Review by orlangur — CHANGES_REQUESTED** [review]

**Comment by orlangur** [comment]
Closing due to inactivity. Feel free to reach me out anytime later if you wish to continue work on this pull request and it will be reopened.
Thanks for collaboration!

**Pull request closed without merging**